secure/storage/vault/src/lib.rs

```diff
@@ -166,6 +202,75 @@ struct ReadSecretData {
     data: HashMap<String, String>,
 }
 
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct ReadPolicyResponse {
+    pub name: String,
+    pub rules: Policy,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+pub struct Policy {
+    #[serde(skip)]
+    internal_rules: PolicyPaths,
+    rules: String,
+}
+
+impl Policy {
+    pub fn new() -> Self {
+        Self {
+            internal_rules: PolicyPaths {
+                path: HashMap::new(),
+            },
+            rules: "".to_string(),
+        }
+    }
+
+    pub fn add_policy(&mut self, path: &str, capabilities: Vec<Capability>) {
+        let path_policy = PathPolicy { capabilities };
+        self.internal_rules
+            .path
+            .insert(path.to_string(), path_policy);
+    }
+
+    pub fn from_json_str(input: &str) -> Result<Self, Error> {
+        let mut policy: Policy = serde_json::from_str(input)?;
+        policy.internal_rules = serde_json::from_str(&policy.rules)?;
+        Ok(policy)
+    }
+
+    pub fn to_json(&self) -> Result<serde_json::Value, Error> {
```

Comment:
this and above feels better as TryFrom

Reply:
This was pretty awful. TryFrom consumes the value, so I need clone. Implementing a proper serializer / deserializer is a ton of boiler plate code since it needs to be generic. The reality is that this is just a json blob hack, so it doesn't make sense to be anything but specific. I implemented the TryFrom, which for whatever reason required an explicit TryInto despite what the docs say. I would argue there's no really attractive solution for this interesting data structure.